ts/src/profiler.ts

```diff
@@ -13,47 +13,292 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import * as delay from 'delay';
+import * as http from 'http';
 import * as path from 'path';
+import * as pify from 'pify';
+import * as zlib from 'zlib';
 
-import {AuthenticationConfig, Common, Logger, Service, ServiceConfig} from '../third_party/types/common-types';
-
-import {Config, ProfilerConfig} from './config';
+import {perftools} from '../../proto/profile';
+import {AuthenticationConfig, Common, Logger, Service, ServiceConfig, ServiceObject, ServiceObjectConfig} from '../third_party/types/common-types';
+import {ProfilerConfig} from './config';
 import {HeapProfiler} from './profilers/heap-profiler';
 import {TimeProfiler} from './profilers/time-profiler';
 
+export const common: Common = require('@google-cloud/common');
 const pjson = require('../../package.json');
-const common: Common = require('@google-cloud/common');
+const API = 'https://cloudprofiler.googleapis.com/v2';
+const gzip = pify(zlib.gzip);
+
+enum ProfileTypes {
+  Wall = 'WALL',
+  Heap = 'HEAP'
+}
+
+/**
+ * Returns true if http status code indicates an error.
+ */
+function isErrorResponseCode(code: number) {
+  return code < 200 || code >= 300;
+}
 
-// TODO: finish implementing Profiler.
-// TODO: add stop() method to stop profiling.
-export class Profiler {
-  config: ProfilerConfig;
-  logger: Logger;
-  service: Service;
+/**
+ * Returns true if http status code indicates request should be retried.
+ */
+function isRetriableResponseCode(code: number) {
```

Comment:
isRetriableStatusCode or isRetriableResponseStatus (saying 'status' is important since that's the error domain here)

Reply:
Done.